GafferImage layer support

johnhaddon opened the pull request:
Following on from yesterday's #2018, this should complete support for multiple image layers (AOVs in rendering terms) in GafferImage. It replaces the old ChannelMaskPlug with simple StringPlugs and `StringAlgo::matchMultiple()` and uses a revamped ChannelMaskPlugValueWidget to help users construct channel masks involving arbitrary channels. It then updates the ImageView so that any layer can be chosen within the viewer, and so that the appropriate display transform is applied to all RGB channels regardless of layer.

I've been fairly gung-ho in adding breaking changes wherever I felt they made most sense for the future (e.g. ImageWriter and DeleteChannels channels defaults). I've tried to keep these in their own commit though in case anyone has concerns (@HughMacdonald?, @lucienfostier?) and would like them delayed.

andrewkaufman commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -705,7 +705,7 @@ ImageWriter::ImageWriter( const std::string &name )
 	addChild( new ImagePlug( "in" ) );
 	addChild( new StringPlug( "fileName" ) );
-	addChild( new StringPlug( "channels", Gaffer::Plug::In, "[RGB]" ) );
+	addChild( new StringPlug( "channels", Gaffer::Plug::In, "*" ) );
```
This seems a much better default, so id be happy to accept the breaking change. I've found it quite fiddly to have to remember to add A manually thus far.

johnhaddon replied:
Cool. I think the original default was conservative because OIIO would fail to write things like JPEGs if we threw too many channels at it, but for a while now we've been testing `out->supports( "channels" )` and `out->supports( "alpha" )` to avoid those problems anyway.

danieldresser commented:
This sounds like exactly what I need.  I guess it's big enough that I should probably have a play with it - I'd be happy to do that, but what's up with the compile failure on Travis?  Looks like one node just didn't get updated?

johnhaddon commented:
I've pushed a fix for the Travis compilation failure. It was still compiling at my end because the install process doesn't remove old headers, so I still had ChannelMaskPlug.h on my machine.

Would be great if you could give this a whirl...

andrewkaufman commented:
needs a rebase

andrewkaufman commented on startup/gui/ocio.py:
```diff
@@ -100,6 +100,7 @@ def __plugSet( plug ) :
 def __displayTransformCreator( name ) :
 
 	result = GafferImage.DisplayTransform()
+	result["channels"].setValue( "*[RGB]" )
```
I don't quite remember how all this works, but this one needs a different default to the one below?

lucienfostier replied:
would this apply a display transform to auxillary channels?
I think we dont want to display something like Z or alpha with a transform.

andrewkaufman replied:
I believe it only applies to RGB channels of arbitrary layers. So. R, G, B, diffuse.R, diffuse.G, diffuse.B would all be affected but Z, A, whatever.Z, whatever.A would not be affected.

johnhaddon replied:
Thanks for spotting that mismatched channel mask - fixed in c80592b. @lucienfostier, yes, we only apply the display transforms to RGB channels, not auxiliary channels.

danieldresser commented:
I'm currently running with this, and it's looking fairly good.

I've only seen two actual issues so far, and one may be causing the other:
* sometimes the layer selector doesn't update while switching nodes
* I get errors like this:
```
Traceback (most recent call last):
  File "/home/danield/apps/gaffer/0.32.0.0dev/cent7.x86_64/cortex/9/gaffer/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/home/danield/apps/gaffer/0.32.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Menu.py", line 188, in __show
    self.__build( self._qtWidget() )
  File "/home/danield/apps/gaffer/0.32.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferUI/Menu.py", line 244, in __build
    definition = definition()
  File "/home/danield/apps/gaffer/0.32.0.0dev/cent7.x86_64/cortex/9/gaffer/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/home/danield/apps/gaffer/0.32.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferImageUI/RGBAChannelsPlugValueWidget.py", line 85, in __menuDefinition
    channelNames = image["channelNames"].getValue()
RuntimeError: Exception : Invalid image file "/data/jobs/POW/sequences/rnd/shots/arnold/3dData/renders/noiseDebug/quality10defireflied/beauty/noiseDebug.quality10defireflied.beauty.00-1.exr"
```
It appears that RGBAChannelsPlugValueWidget is sometimes evaluating the scene at frame 1 ( it's a known issue with my network that I access two frames before with a TimeWarp, and am not currently dealing correctly with the edge of the sequence.


One thing that is either a bug or very confusing:  the layer selection dropdown in the viewer is a checklist?  I would have expected to just be selecting one layer, I'm not sure what it means to check multiple layers?

Things that don't necessarily need to be addressed as part of this PR, but would be nice at some point:

* The channel isolate widget in the viewer always shows the channels RGBA to choose from, it seems like it would be nice if this was dynamic?  If I'm viewing an image that has one channel named "foo.foo", and I've selected layer foo, should it indicate that?  Or if for some reason I only have a red and blue channel but no green?

* There's currently no way to actually create arbitrary named channels in the UI?  I have to use the script editor to set the out channel on a Shuffle quite frequently, presumably this UI will be updated at some point?

* Now that we've got this slick channel filtering UI, would it make sense to be able to select channels on a Merge node, and leave the rest alone?  This would save me some places where I have to Merge then CopyChannels just the ones I want.


Oh, GitHub was prompting me to fix a merge conflict, so I clicked the little button and fixed it in the GitHub UI, and it added a commit for "Merge branch 'master' into viewerLayerSupport" - it was a trivial fix, but hopefully this doesn't confuse the history.

Also, I restarted the tests, since it look like it got a random failure.  There was an issue with renderdl on Mac, also a nasty one that looked like it was a segfault in some sort of graph traversal for dirty propagation.  It seems like this has started happening for the last while, it's sporadic enough that it may be tough to track down what introduced it.  I couldn't immediately find the logs of the fail on this PR, I guess since I restarted it, but there's an example here:
https://travis-ci.org/GafferHQ/gaffer/jobs/216461520
That's after I merged your MedianFilter PR ( the tests passed on the PR, but failed after merging ).

From my point of view, this feels very close to mergeable.  It definitely seems like a step forward, and I'm not worried about backwards compatibility too much yet.

johnhaddon commented:
> It appears that RGBAChannelsPlugValueWidget is sometimes evaluating the scene at frame 1 ( it's a known issue with my network that I access two frames before with a TimeWarp, and am not currently dealing correctly with the edge of the sequence.

Thanks for catching that - fixed in 78aa7b7. In 	f6e1c35 I've also made it fail more gracefully for cases where there is a genuine error even when using the appropriate context.

> sometimes the layer selector doesn't update while switching nodes

I'm not sure if this is related to the above or not, because I'm not sure exactly what you're referring to. The bug above would mean that the menu would error and potentially show the wrong thing - this is now fixed. But I never intended for the actual layer selection to automatically change when switching from an RGBA image to a diffuse.RGBA image for example, so hopefully that's not what you're referring to.

> One thing that is either a bug or very confusing: the layer selection dropdown in the viewer is a checklist? I would have expected to just be selecting one layer, I'm not sure what it means to check multiple layers?

You can't select multiple layers - selecting a different one unselects the previous one. Having the current selection checked is a pretty standard UI idiom though isn't it? We do it all over in Gaffer (presets menus for example), and out of paranoia I just checked and found the same thing all over my native macOS UIs as well. What's making it confusing in this instance?

> Things that don't necessarily need to be addressed as part of this PR, but would be nice at some point:

Those all seem reasonable, but I'd rather they we did leave them for another time. I think we've got a pretty decent first pass together here. Time to move on and build the Catalogue this makes possible...

> Oh, GitHub was prompting me to fix a merge conflict, so I clicked the little button and fixed it in the GitHub UI, and it added a commit for "Merge branch 'master' into viewerLayerSupport" - it was a trivial fix, but hopefully this doesn't confuse the history.

Ugh. I can't abide the merge-from-master-to-branch-then-merge-back-to-master thing - is there anything good about it in comparison to a rebase? I had to add more commits in any case, so I've just rebased.

> Also, I restarted the tests, since it look like it got a random failure. It seems like this has started happening for the last while, it's sporadic enough that it may be tough to track down what introduced it.

Yeah, these have been plaguing us for a little while, but I've been unable to get anything to reproduce outside of Travis.

> From my point of view, this feels very close to mergeable. It definitely seems like a step forward, and I'm not worried about backwards compatibility too much yet.

Great - if my fixes do the job for you and noone objects to the compatibility changes, could you go ahead?

danieldresser commented:
> sometimes the layer selector doesn't update while switching nodes

Hmm, whatever it was, seems to be fixed now.

> You can't select multiple layers - selecting a different one unselects the previous one.

I was sure last night that I saw multiple selected, but I was getting really confused by some out of date UIs that I think were related to exceptions that you've fixed now.  Either it was because of that or I was hallucinating, but it seems to be working fine now.

It's late enough in the day that I'm not really feeling like taking the initiative to push the big green merge button, but I'm now running with this, and it's working fine for me, so from my point of view it's good to go.

danieldresser commented:
Forgot to mention this before leaving, but Andrew said he had looked through the code, and thought it was good to merge, so yeah, go ahead merge, or I'll do it first thing tomorrow morning.

johnhaddon commented:
Great - thanks for the testing.
